src/website/shared/models/linkage.py

```diff
@@ -68,3 +72,55 @@ class DerivationClusterProposalLink(models.Model):
     # TODO: how to design the integrity here?
     # we probably want to add a fancy check here.
     provenance_flags = models.IntegerField()
+
+
+@receiver(post_save, sender=CVEDerivationClusterProposal)
+@receiver(m2m_changed, sender=DerivationClusterProposalLink)
+def sync_proposals_updated_at(
+    sender: CVEDerivationClusterProposal | DerivationClusterProposalLink,
+    instance: CVEDerivationClusterProposal,
+    **kwargs: dict[str, Any],
+) -> None:
+    """Keep proposal's `updated_at` field in sync with last relevant update.
+
+    From this function we can import `DerivationClusterProposalLinkEvent`
+    without import errors. When we try to import it at the module level,
+    there is an error about `shared.models` being partially initialized.
+    This is happening because the `Event` models get dynamically added by
+    the `@pghistory.track` decorators. The alternatives to having the import
+    in the function are:
+        - Creating the Event model explicitly.
+        - Using `instance.derivations.through.pgh_event_model` where
+          `DerivationClusterProposalLinkEvent` is used. However, that
+          expression is less readable.
+
+    Keeping `updated_at` in sync would be done more safely with Postgres triggers.
+    However, introducing them for this case would involve hardcoding table names.
+    Consider changing this logic to triggers once upstream's proposal to support
+    Django templating to generate triggers that involve different tables
+    (such as this denormalization sync):
+        - https://github.com/Opus10/django-pgtrigger/discussions/165
+    """
+    from shared.models import DerivationClusterProposalLinkEvent  # type: ignore
+
+    last_inserted_timestamp = instance.updated_at
+
+    if sender == CVEDerivationClusterProposal:
+        if kwargs["update_fields"] and "status" in kwargs["update_fields"]:
+            last_inserted_timestamp = instance.status_events.order_by(  # type: ignore
+                "-pgh_created_at"
+            ).values_list("pgh_created_at", flat=True)[0]
+
+    else:
+        if kwargs["action"] and kwargs["action"] == "post_remove":
+            last_inserted_timestamp = (
+                DerivationClusterProposalLinkEvent.objects.filter(
+                    proposal_id=instance.pk
+                )
+                .order_by("-pgh_created_at")
+                .values_list("pgh_created_at", flat=True)[0]
+            )
+
+    if instance.updated_at is None or last_inserted_timestamp > instance.updated_at:  # type: ignore
+        instance.updated_at = last_inserted_timestamp
+        instance.save()
```

Comment:
Minimally, this should be protected against infinite looping by asserting that this save won't trigger further signal receives.

Reply:
This is what is achieved by having the check on "status" in "updated_fields".